File: 2023-03/python/part1.py
import re

NUMBER_REGEX = re.compile(r"(\d+)")
SYMBOL_REGEX = re.compile(r"[^\d\.\s]")


def get_range(vec, imin, imax, inclusive=False):
    return vec[max(0, imin): min(len(vec) - 1, imax) + 1 if inclusive else min(len(vec), imax)]


def is_symbol_adjacent(lines, line_idx, span):
    return any(map(lambda line: SYMBOL_REGEX.search(get_range(line, span[0] - 1, span[1] + 1)),
                   get_range(lines, line_idx - 1, line_idx + 1, inclusive=True)))


def run(file):
    lines = file.readlines()
    result = 0
    for i, line in enumerate(lines):
        for c in NUMBER_REGEX.finditer(line):
            if is_symbol_adjacent(lines, i, c.span()):
                result += int(c.group())
    return result


if __name__ == "__main__":
    with open("../input.txt", 'r') as f:
        print(run(f))

File: 2023-03/rust/part1.rs
pub mod get_relaxed;
use crate::get_relaxed::GetRelaxed;

use lazy_static::lazy_static;
use regex::Regex;
use std::fs::read_to_string;
use std::ops::Range;

lazy_static! {
    static ref NUMBER_REGEX: Regex = Regex::new(r"\d+").unwrap();
    static ref SYMBOL_REGEX: Regex = Regex::new(r"[^\d\.\s]").unwrap();
}

fn is_symbol_adjacent(lines: &Vec<&str>, line_idx: usize, span: Range<usize>) -> bool {
    lines
        .get_relaxed(line_idx as i32 - 1..=line_idx as i32 + 1)
        .iter()
        .any(|line| {
            SYMBOL_REGEX.is_match(line.get_relaxed(span.start as i32 - 1..span.end as i32 + 1))
        })
}

fn run(filename: &str) -> i32 {
    let file_content = read_to_string(filename).unwrap();
    let lines: Vec<&str> = file_content.lines().collect();
    lines
        .iter()
        .enumerate()
        .map(|(idx, line)| {
            NUMBER_REGEX
                .captures_iter(line)
                .map(|c| c.get(0).unwrap())
                .filter(|m| is_symbol_adjacent(&lines, idx, m.range()))
                .map(|m| m.as_str().parse::<i32>().unwrap())
                .sum::<i32>()
        })
        .sum()
}

fn main() {
    let result = run("../input.txt");
    println!("{result}");
}
